HPCC-29657 Produce aggregate stats (e.g. spill, cost) whilst a job is running

shamser opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

shamser committed:
HPCC-29657 Produce aggregate stats (e.g. spill, cost) whilst a job is running

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.com>

shamser commented:
Note: Work in progress hence draft PR

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -1825,7 +1813,7 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
     virtual StringBuffer & getFullScope(StringBuffer & str) const override
     {
-        if (parent)
+        if (parent && queryScopeType()!=SSTworkflow)
```
better test would be parent->queryscopeType() != SSTglobal.  (Would work with compile stages as well.)

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -1988,9 +2000,58 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
             cur.visit(visitor);
     }
 
+    virtual bool refreshAggregates(std::vector<StatisticKind> & aggregateKinds, std::vector<unsigned __int64> & totals) override
```
Would a CRuntimeStatisticCollection (instead of aggregateKinds, totals) simplify this code?  The dual arrays seem to implement the same idea.

shamser replied:
CRuntimeStatisticCollection cannot store the variants.  When the stats are deserialized, we'd have to ignore all the variants.  We could ignore all the variants.  Losing the variants would mean that we could not have a single in-memory object for the stats as we'd need one for aggregation (GlobalStatisticCollection ) and another for serialization(CStatisticGatherer).  We would also not be able to extend the functionality to use GlobalStatisticCollection to cache runtime stats and allow periodic updates to dali.

shamser replied:
Also, not that a new CRuntimeStatisticCollection will need to be created for each scope and each one will have a StatisticMapping. With the current implementation, a single vector with the StatisticKinds is used as it traverses the entire StatisticsCollection.  I think using CRuntimeStatisticCollection will be slower and use more memory.

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -2108,6 +2169,51 @@ class CRootStatisticCollection : public CStatisticCollection
+                StatsScopeId id;
+                id.deserialize(serialized, version);
+                IStatisticCollection * collection = ensureSubScope(id, true);
+                collection->deserialize(serialized, version);
```
Optimization for later.  Indicate to deserialize that it does not need to deserialize beyond a subgraph.  Would avoid all the activities being deserialized.
That option might be usable by other stats iterators.

shamser replied:
Noted.  I'll avoid deserialize unnecessary scopes.

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -2108,6 +2169,51 @@ class CRootStatisticCollection : public CStatisticCollection
+class GlobalStatisticCollection : public CStatisticCollection
+{
+public:
+    GlobalStatisticCollection(IPropertyTree * root) : CStatisticCollection(nullptr, globalScopeId)
```
This code belong in workunit.cpp, rather than jlib, because it is specific to the way stats are represented in workunits.  It should be possible to implement using the public interface for the stats collection.

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -1988,9 +2000,58 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
+                while (totalIter != totals.end())
+                {
+                    StatsMergeAction mergeAction = queryMergeMode(*kindIter);
+                    updateStatistic(*kindIter, *subTotalIter, mergeAction);
```
I think this should probably replace rather than merge - since this is the new aggregated value.

shamser replied:
I was thinking StatsMergeAction would help to keep track of peaks, such as PeakSpillSize. What do you think?  Is simple aggregation enough (replace action) or should we try to handle other types such as peaks?

ghalliday reviewed (commented):
@shamser looks good.  A few comments.

The main question in my mind is where should stats that are going to be aggregated be stored?  This change has moved them from the global stats to the subgraph.  I think that makes logical sense, but I don't know what implications that has for efficiency.  (Logically the execution times should also move, but that would have a performance impact.)

The other question in my mind is what is the minimal set of changes which would allow this change to be merged?


The other thought is

shamser commented:
> I think that makes logical sense, but I don't know what implications that has for efficiency. (Logically the execution times should also move, but that would have a performance impact.)
> 
> The other question in my mind is what is the minimal set of changes which would allow this change to be merged?
> 
> The other thought is

I didn't move execution times because there was some special aggregation code that handled the case where workflow and hthor was on the same cpu. Also, it wasn't the case the aggregation of execution time at sub-graph level should be the same as the execution of the parent scopes.  For example, there was also small gaps between subgraphs which would be lost if we simply aggregated the subgraphs.

I did move StSizeSpillFile, StSizeGraphSpill and StCostFileAccess which I thought wouldn't impact performance too much.  Are there instances that you are aware of that would be impacted by these changes?

shamser commented:
@ghalliday I pushed 2 new commits:
1) Relocate GlobalStatisticCollection to workunit.cpp
Prepare dynamic stats tracking by GlobalStatisticCollection
Remove unnecessary calls to updateAggregation
2) Track active graph stats in GlobalStatisticCollection (meaning that the entire graph does not need to loaded from dali for stats aggregration)

With the second commit, I eliminated the loading of stats from Dali for aggregation in Thor.  I'm making the similar changes in hThor.

shamser commented on thorlcr/master/thdemonserver.cpp:
```diff
@@ -69,6 +69,10 @@ class DeMonServer : public CSimpleInterface, implements IDeMonServer
+                cost_type costDiskAccess = graph->getDiskAccessCost();
+                if (costDiskAccess)
+                    stats.addStatistic(StCostFileAccess, costDiskAccess);
+
```
This should really move to MasterGraph::getStats()

jakesmith commented:
@shamser - please see inlne comments, and comments below.

If I understand it correctly, the current implementation does:
1) load stats from subgraphs only (i.e. does not load previously published aggregates that have been pushed to workunit <Statistics> at the moment).

2) use statsMaxDepth to skip over intermediate levels *until* it reaches depth 3 (this constant isn't very clear, could it use a StatsScopeId instead?).
The comment says ignore stats below subgraph, but afaics, it doesn't append until it gets to 3, then adds everything below?

3) use missingScopesOnly to check if a subgraph has already been deserialized (already a child collection).


Goals:
- we want to avoid recalculating aggregates if nothing has changed.
- load existing stats if they will be needed to upgrade an incomplete parent aggregate
- avoid loading stats and totals below a level, unless a parent is marked dirty.

As it stands, when a new thor instance picks up a running workunit, let's say to run graph100, I think it will:
- load - loads all stats from all subgraphs from all graphs, in all workflow items.
- updateAggregates - call will recalculate all aggregates for all levels, but, not publish them if they are equal to what is already published.
- missingScopesOnly avoids re-fetching/re-deserializing the binary blobs for subgraphs if the current Thor instance has already done so, but doesn't help if the Thor instance hasn't seen this workunit before (or it had, but the last workunit was different).


As it stands, there would be too much of a performance hit on medium to big size jobs, unless you had just 1 Thor instance that worked on the whole job from start to finish.

To overcome that, as we've discussed and as you've alluded to in one of your comments, we need a mechanism which tracks what levels are dirty.
And, the implementation needs to load existing aggregates at startup too, and avoid loading nested stats. unless a level is dirty (i.e. the totals for it need recalculating because it's not yet done).
Published aggregates will change and need recalculating until the level is done, e.g. the aggregates in WF1 will need to be constantly recalculated until all the graphs under it are complete, i.e. when WF1 is complete.
Same for graphs, e.g. aggregates for graph59 will need recalculating until all its subgraphs are done. NB: in practice atm, there won't be a situation where the engines start processing a graph with some subgraphs done and some not - but that could change, e.g. with checkpointing. However, the logic is the same for dirty graphs/subgraphs as it is for workflow items/graphs.

I think this can be achieved by adding a dirty flag (not sure whether best under <GraphProgress> or <Statistics>) to each level that is active, e.g. when WF2 starts, it is marked dirty, and the flag is only cleared when it completes.
As GlobalStatisticCollection::load walks the stats, it can stop descending to load lower levels if there is no dirty flag. e.g. when loading the stats for a job running WF2, WF1 totals are loaded, but since it has no dirty flag, none of the graph aggregates, or subgraph stats below WF1 need loading (or need processing later when it comes to updateAggregates).
NB: there's probably little point in trying to avoid any aggregates under <Statistics> from being loaded, and this will be particularly true if we move them to a single serialized blob when the whole blob will need writing back, but it can use this logic to avoid descending into graphs/subgraphs that won't be needed when it's walking <GraphProgress>.

updateAggregates needs to recalculate and push all aggregates for any dirty layer (the top/global level is implicitly dirty if the job is running anything).

In practice that means, that when it's e.g. running graph100 in wf5 and the job lands on a new Thor instance, global+wf5+graph100 should be the only things that are marked dirty.
load() will need to load all existing aggregates for wf1, wf2, wf3, wf4, but none of their child graphs stats, and because wf5 is dirty, load all aggregates for graph1-graph99, but none of their children because graph1-graph99 are not dirty.
And because graph100 is dirty, logically it should load all of the published subgraph aggregates, *but*, since a graph is re-entrant at the moment (can't run half of it then resume on another instance), it is always going to redo the complete graph. Which means that for now, we should consider all subgraph aggregates and stats as missing on load I think (for the purposes of calculating new aggregates) - they are of no use. That will change if we start supporting checkpointing again, so we should comment clearly why we don't bother loading existing subgraph aggregates and stats.

shamser commented on common/workunit/workunit.cpp:
```diff
@@ -2659,83 +2665,114 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+    IStatisticCollection * graphScopeCollection = wfScopeCollection->ensureSubScope(graphScopeId, true);
+    IStatisticCollection * sgScopeCollection = graphScopeCollection->ensureSubScope(sgScopeId, true);
+    if (clearStats)
+        sgScopeCollection->clearStats();
```
After a call to getCollectionForUpdate(), all stats for this scope will be rewritten. Nothing will be lost.

shamser commented on system/jlib/jstats.cpp:
```diff
@@ -1988,9 +2012,69 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
+                std::vector<StatisticKind>::iterator kindIter = aggregateKinds.begin();
+                while (totalIter != totals.end())
+                {
+                    if (updateStatistic(*kindIter, *subTotalIter, StatsMergeReplace))
```
This is StatsMergeReplace because it is updating a single value for this scope.   The value has already been "merged" at the child level.  The loop 2092-2101 generates the subTotal value by using the correct mergeAction (line 2101).  In line 2122, the one subTotal per kind is written to this scope.

shamser commented on system/jlib/jstats.cpp:
```diff
@@ -1988,9 +2012,69 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
+            {
+                Statistic & stat = stats.element(i);
+                StatisticKind kind = stat.queryKind();
+                auto iteratorVec = std::find(aggregateKinds.begin(), aggregateKinds.end(), kind);
```
Are you sure that the order of the statistics is consistent? For example, if a statistic has a value of zero, addStatistic/updateStatistic may not be called so it would not be in stats.

shamser commented on system/jlib/jstats.cpp:
```diff
@@ -1988,9 +2012,69 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
+                if (iteratorVec!=aggregateKinds.end())
+                {
+                    unsigned pos = iteratorVec-aggregateKinds.begin();
+                    totals[pos] += stat.queryValue();
```
Yes, I've changed this.

shamser commented on common/workunit/workunit.cpp:
```diff
@@ -2659,83 +2665,114 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+
+            StatsScopeId childId;
+            childId.deserialize(serialized, version);
+            statsCollection->deserializeChild(childId, serialized, version, 3);
```
The design has changed since this comment was posted.

shamser commented:
> tes. As GlobalStatisticCollection::load walks the stats, it can stop descending to load lower levels if there is no dirty flag. e.g. when loading the stats for a job running WF2, WF1 totals are loaded, but since it has no dirty flag, none of the graph aggregates, or subgraph stats below WF1 need loading (or need processing later when it comes to updateAggregates). NB: there's probably little point in trying to avoid any aggregates under from being loaded, and this will be particularly true if we move them to a single serialized blob when the whole blob will need writing back, but it can use this logic to avoid descending into graphs/subgraphs that won't be needed when it's walking .
> 
> updateAggregates needs to recalculate and push all ag

I've modified the design since this comment was posted.  I was actually testing the changes before I saw these code review comments.  Although the design is different from what you've suggested, I feel that it achieves that efficiency goal. Please can you have a look at the PR again. @jakesmith

jakesmith commented on system/jlib/jstats.cpp:
```diff
@@ -1974,43 +1973,52 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
+            curScope = curScope->querySubScope(scopeItem);
             if (!curScope)
-                return nullptr;
+                break;
```
this doesn't seem right.
This break; (and falling through to return curScope) means as it's traversing the StatsScopeId list, if one isn't found it will return the IStatisticCollection level of wherever it happened to get to when it stopped.

It seemed right as it was, to return null if it can't satisfy the full query.

shamser replied:
It only breaks if curScope is null, so if it break here the return will be null.  I could return a null pointer here but I don't think it makes much difference, does  it?

(I did it this way because when I was debugging, it was easier to add debug just before a single return line)

jakesmith replied:
I misread

jakesmith commented on thorlcr/master/thdemonserver.cpp:
```diff
@@ -152,6 +153,8 @@ class DeMonServer : public CSimpleInterface, implements IDeMonServer
                     unsigned startTime = graphStarts.item(g2);
                     reportStatus(wu, graph, startTime, finished, success);
+                    // Prune subgraph descendant stats as they have been serialized and no longer needed.
+                    globalStatsCollection.pruneSubGraphDescendants(wfid, graphName, graph.queryGraphId());
```
Also, whilst true these aren't needed in globalStatsCollection, they will be repopulated repeatedly on every call to reportActiveGraphs.

I think it would better to clear once, in endGraph()

jakesmith commented on thorlcr/master/thdemonserver.cpp:
```diff
@@ -172,10 +175,12 @@ class DeMonServer : public CSimpleInterface, implements IDeMonServer
             }
-
+            // Prune subgraph descendant stats as they have been serialized and no longer needed.
+            globalStatsCollection.pruneSubGraphDescendants(wfid, graphName, graph->queryGraphId());
```
see above comment, re. clearing once in endGraph()

jakesmith commented on common/workunit/workunit.cpp:
```diff
@@ -2657,85 +2660,187 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+    }
+
+    loadGlobalAggregates(workunit);
+    Owned<IPropertyTree> root = getWUGraphProgress(wuid, true);
```
The normal expected usage of GlobalStatisticCollection::load, is that the graphName provided is graph that either hasn't been run, or is being rerun and it's stats. previously clearer (if not cleared, the aggregates would be confusing..).

I think should elude to that here, to say that there are generally not expected to be any subgraph stats from graphName here.

I wonder whether the engines should pass in graphName=nullptr , to explicitly tell it not to even bother to try to load. Which would save a call + lock to dali on /GraphProgress which isn't going to be used.

jakesmith commented on thorlcr/master/thgraphmanager.cpp:
```diff
@@ -1134,7 +1136,9 @@ bool CJobManager::executeGraph(IConstWorkUnit &workunit, const char *graphName,
         cost_type cost = money2cost_type(calculateThorCost(nanoToMilli(graphTimeNs), numberOfMachines));
         if (cost)
             wu->setStatistic(queryStatisticsComponentType(), queryStatisticsComponentName(), SSTgraph, graphScope, StCostExecute, NULL, cost, 1, 0, StatsMergeReplace);
```
will these graph CostExecute stats. be aggregated at the workflow level, and global level?

shamser replied:
The workflow level cost includes not only all the subgraph costs but also workflow engine cost.  There is special code to handle this calculation: when calculating workflow level cost, it adds the additional workflow engine cost to the subgraph cost aggregates.  Unfortunately, this also means that the graph aggregates cannot be calculated using the new method, as it would override workflow cost calculated with the special cost.

shamser replied:
I have looked into making use of the global stats collection object to handle CostExecute.  It may make the code cleaner.  However, this is not a trivial change and, probably, isn't worth doing in this PR.  (The workflow engine costs complicates things too much).

jakesmith commented on common/workunit/workunit.cpp:
```diff
@@ -2657,85 +2660,187 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+            }
         }
+    };
+    if (refreshAggregates()) // Only serialize if the aggregates has changed
```
trivial: has->have

jakesmith commented on common/workunit/workunit.cpp:
```diff
@@ -2657,85 +2660,187 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+                {
+                    StatisticKind kind = aggregateKindsMapping.getKind(i);
+                    stat_type value;
+                    if (cur.getStatistic(kind, value) && value)
```
you do sometimes want to set if 0, it depends, should be using includeStatisticIfZero(kind) I think.

jakesmith commented on common/workunit/workunit.cpp:
```diff
@@ -2657,85 +2660,187 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+    if (refreshAggregates()) // Only serialize if the aggregates has changed
+    {
+        StatisticsAggregatesWriter statsAggregatorWriter(wu, aggregateKindsMapping);
+        statsCollection->visit(statsAggregatorWriter);
```
related to the includeStatisticIfZero comment, should this be doing something like : ?

```
Owned<IStatisticGatherer> globalStaatsGatherer = createGlobalStatisticGatherer(wu);
statsCollection->recordStatistics(globalStatsGatherer);
```

CRuntimeStatisticCollection::recordStatistics handles things like includeStatisticIfZero, merge mode and measure conversion.
createGlobalStatisticGatherer builds up a scope as it's descending.

shamser replied:
statsAggregatorWriter is filtering out the required stats and writing out only those stats. It filters out some global stats that have been loaded in global stats such as dfu stats, compile stats etc.  Stats at the subgraph level should also be ignored (not written to global wu stats).

Also, IStatisticCollection/CStatisticCollection doesn't have a recordStatistics function.

jakesmith reviewed (commented):
@shamser - looks good and close, there's a couple of leaks that needs fixing, some trivial things and a couple of questions, but I suspect those could be done in a subsequent PR.

Please address the leak/trivials for now.

shamser commented on system/jlib/jstats.cpp:
```diff
@@ -2183,7 +2191,8 @@ StringBuffer &CStatisticCollection::toXML(StringBuffer &out) const
         iter.query().toXML(out);
-    out.append("</Scope>\n");
+    out.append("</Scope> // Scope id=\"");
+    id.getScopeText(out).append("\"\n");
```
yes, extra debugging.  It's not used for any purpose other than debugging, so no harm in leaving this in.

ghalliday replied:
Except it isn't valid xml any more.  The information is available in the opening tag why duplicate it here?

jakesmith replied:
Unresolving this comment - missed fact that it is not a valid XML comment syntax, and Gavin's other Q re. duplicate info.

jakesmith commented on ecl/eclagent/eclagent.ipp:
```diff
@@ -705,6 +714,15 @@ public:
+    virtual IWUGraphStats *updateStats(StatisticCreatorType creatorType, const char * creator, unsigned activeWfid, const char *graphName, unsigned subgraph) override
+    {
+        Owned<IStatisticCollection> sgCollection = globalStats.getCollectionForUpdate(creatorType, creator, activeWfid, graphName, subgraph, true); // true=>clear existing stats
+        return wuRead->updateStats (graphName, creatorType, creator, activeWfid, subgraph, false, sgCollection);
```
space after updateStats before (

jakesmith commented on thorlcr/master/thdemonserver.cpp:
```diff
@@ -301,6 +301,18 @@ class DeMonServer : public CSimpleInterface, implements IDeMonServer
         reportActiveGraphs(true, false);
+        if (activeGraphs.ordinality())
+        {
+
```
extra blank line

jakesmith reviewed (commented):
@shamser - please see new comments.

Mostly cosmetic and 1 question re. isDirty handling:  https://github.com/hpcc-systems/HPCC-Platform/pull/17786/files#r1377617622

shamser commented on system/jlib/jstats.cpp:
```diff
@@ -1994,12 +2098,82 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
+    {
+        const StatisticsMapping & mapping = totals.queryMapping();
+        bool updated = false;
+        if (queryScopeType()==SSTsubgraph || isDirty==false)
```
I agree: checking isDirty is sufficient.  I had the subgraph test because the isDirty flag was implemented later and I didn't reassess the need for the subgraph test.

I modified the code so that if isDirty is false at the global level, the operation is a nop.

shamser replied:
Ignore the previous comment.  The unconditional sg test is required.  The first block (if condition true) returns the totals at this level.  The second block (if condition false) descends to a lower scope level to get totals from lower level.

It should return totals from the current level under 2 circumstances
1) isDirty==false: the stats at this level and descendant levels haven't changed, so it is ok to use the aggregates from this level
2) it is at subgraph scope:  this condition is required because even if the dirty flag is true, aggregates should only be generated from subgraph level.  As per the agreed requirements, the aggregates generated automatically should not be generated from activities->subgraph.  (At a later date, if it is determined that aggregates should be generated from aggregates, this condition may be removed but it is beyond the scope of this jira for this change now.)

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -1782,33 +1785,35 @@ class SortedCollectionIterator : public ArrayIIteratorOf<IArrayOf<IStatisticColl
 
-    CStatisticCollection(CStatisticCollection * _parent, MemoryBuffer & in, unsigned version) : parent(_parent)
+    // deserialize without storing the stats
+    void deserializeNoStats(MemoryBuffer & in, unsigned version)
```
style: In this file/class private members go after the public members

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -1747,11 +1750,11 @@ enum
 
 class CStatisticCollection;
-static CStatisticCollection * deserializeCollection(CStatisticCollection * parent, MemoryBuffer & in, unsigned version);
+static IStatisticCollection * deserializeCollection(IStatisticCollection * parent, MemoryBuffer & in, unsigned version);
```
Why these changes?

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -1831,10 +1836,11 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
     virtual StringBuffer & getFullScope(StringBuffer & str) const override
     {
-        if (parent)
+        if (parent && queryScopeType()!=SSTglobal)
```
Why would a global scope have a parent?

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -1862,6 +1868,24 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
+        else
+        {
+            StatsScopeId childScopeId;
+            if (!childScopeId.setScopeText(scope, &scope) || (*scope!=':' && *scope!='\0'))
```
Clearer to use different variables for input and output.  E.g. next.

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -1862,6 +1868,24 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
         }
         return false;
     }
+    virtual bool setStatistic(const char *scope, StatisticKind kind, unsigned __int64 & value) override
```
Doesn't really need to be a member function - could be implemented using the existing public interface.

shamser replied:
It is used from workunit.cpp by the aggregate loader.  It could be done using the existing member functions but it would be cumbersome to add code to the aggregate loader to split the scopes and navigate down the tree.

ghalliday replied:
I agree it should be in a function.  The comment is that it could be a global function - which would reduce the number of functions in the interface.  Debatable style wise, but was connected to the large number of new functions in the interface.

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -1967,6 +2034,43 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
+        }
+        else
+        {
+            deserializeNoStats(in, version);
```
This seems a shame.  Can it be avoided for the common case i.e. where you are only reading the subgraph stats which are the root for hthor/thor graphs.

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -1994,12 +2098,95 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
+        {
+            // descend down to lower level to obtain totals required for aggregates and then aggregate
+            CRuntimeStatisticCollection childTotals(mapping);
+            Owned<IBitSet> childTotalUpdated = createBitSet(mapping.numStatistics());
```
This logic does not seem correct.  total will only updated with any values that have changed - but in the not-dirty case all values are updated.

jakesmith replied:
Agree, I don't follow the logic (related to previous comment here: https://github.com/hpcc-systems/HPCC-Platform/pull/17786#discussion_r1377617622).

The true condition will be followed unconditionally for all subgraphs (regardless of whether isDirty = true or false), which looks like it means everything above every subgraph will also be updated.
Maybe

jakesmith replied:
I think it achieves the goal, but am still finding how it is doing it a bit confusing.
What I'd expect I think, is for each level to do:
+ if at level and dirty, do nothing
+ If at level and is subgraph, do nothing.
+ If dirty, recurse down each scope, and then merge the child values into new totals at current level.
+ When finished recursing down each child scope and merging into new totals, update what has changed into current level stats, and push totals (via callback if necessasry).


I had a stab at what that would look like, and I also think it should be done without adding the two quite implementation orientated 'refreshAggregates' methods to IStatisticCollection : 

```
bool refreshAggregates(IStatisticCollection &stats, const StatisticsMapping & mapping, AggregateUpdatedCallBackFunc & fWhenAggregateUpdated)
{
    if (stats.queryScopeType() == SSTsubgraph) // don't descend any further
        return false;
    if (!stats.isDirty())
        return false;
    bool anyChanges = false;
    CRuntimeStatisticCollection totals(mapping);
    Owned<IBitSet> updated = createBitSet(mapping.numStatistics());
    Owned<IStatisticCollectionIterator> childIter = &stats.getScopes(nullptr, false);
    ForEach(*childIter)
    {
        IStatisticCollection &child = childIter->query();
        refreshAggregates(child, mapping, fWhenAggregateUpdated);

        // merge in child level stats into current level, and mark which aggregate kinds have updated
        unsigned numChildStats = child.getNumStatistics();
        for (unsigned s=0; s<numChildStats; s++)
        {
            StatisticKind kind;
            unsigned __int64 value;
            child.getStatistic(kind, value, s);
            if (kind != (StatisticKind)(kind & StKindMask))
                continue; // ignore variants->not supported by CRuntimeStatisticCollection
            unsigned index = mapping.getIndex(kind);
            if (index != mapping.numStatistics())
            {
                totals.mergeStatistic(kind, value);
                updated->set(index); // NB: may already have been set whilst merging in another child, but that's okay.
            }
        }
    }

    // any updates? Set new totals into current level, and call callback
    unsigned nextUpdateIndex = 0;
    while (true)
    {
        nextUpdateIndex = updated->scan(nextUpdateIndex, true);
        if (NotFound == nextUpdateIndex)
            break;
        
        StatisticKind kind = mapping.getKind(nextUpdateIndex);
        unsigned __int64 value = totals.queryStatisticByIndex(nextUpdateIndex).get();
        if (stats.updateStatistic(kind, value, StatsMergeReplace))
        {
            if (value || includeStatisticIfZero(kind))
            {
                StringBuffer s;
                fWhenAggregateUpdated(stats.getFullScope(s).str(), stats.queryScopeType(), kind, value);
                anyChanges = true;
            }
        }
    }
    stats.clearDirty();
    return anyChanges;
}
```

ghalliday reviewed (commented):
@shamser Broadly the change looks good.  There are two general comments:
- I expected the collection code to stay as it is, and then have a function that merges the subgraph(/graph) stats into the aggregator as a separate call.  Storing the stats in the global stats aggregator seems to be overloading the meaning of that class.
- The public interface to the IStatisticsCollection class expanded by more than I expected.  Some could be implemented as global functions, some indicate logic ins't quite in the right place.

Also a few bits of code where the logic does not look right/should be deleted.

- non-memberMore information is leaked out of the jstats files that I would expect.  is

shamser commented on common/workunit/workunit.cpp:
```diff
@@ -2657,85 +2660,192 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
-    Owned<IConstWUScopeIterator> it = &wu->getScopeIterator(filter);
-    cost_type totalCost = 0;
-    for (it->first(); it->isValid(); )
+    const char * _wuid = workunit.queryWuid();
```
It isn't at the moment.  However, it may be needed later.  For example, when a graph is resumed/restarted, the previous sg stats may be needed.  The costs from previous runs may need to be preserved and aggregated to existing costs.

jakesmith replied:
agree, I think aggregateOnly is only ever going to be true for the foreseeable future, and it would be good to remove this option and simplify the code.

shamser commented on common/workunit/workunit.cpp:
```diff
@@ -183,14 +183,17 @@ void doDescheduleWorkkunit(char const * wuid)
 {
     StatsScopeId graphScopeId;
     verifyex(graphScopeId.setScopeText(_rootScope));
+    StatsScopeId wfScopeId(SSTworkflow,wfid);
```
No it hasn't.  It was previously _called_ "rootScopeId" but it was actually a workflow scope. Line 192 of the original code has "StatsScopeId rootScopeId(SSTworkflow,wfid)".  I changed the name from rootScopeId to workflowScopeId as it was being confused with _rootScope.

shamser commented on common/workunit/workunit.cpp:
```diff
@@ -2657,85 +2660,192 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
     }
+};
+
+GlobalStatisticCollection::GlobalStatisticCollection() : aggregateKindsMapping(aggregateKindStatistics)
```
I was thinking that we'd have a single location for all the stats being gathered and aggregated.      I was trying to avoid having one class to gather and then have another for aggregation - and then having to copy stats between the two.

jakesmith replied:
I think one of the other reasons it was more generalized (not just a aggregator), was that there was a separate future intent for it to manage the gathered stat. collections - and publish them en masse periodically (rather than end of sg as it is now).
That is a separate problem - updating Dali stats too frequently when there are many small quick running subgraphs.

ghalliday replied:
ok.  That could be achieved more simply by saving away a list of blobs to commit, and writing them all in one transaction.

I'm not sure if this is worth changing now - I will think about it, but it does complicate the implementation by combining two different concepts and needing to worry about pruning results etc.  It makes it harder to understand the responsibilities and implementation of the class etc.

shamser commented on common/workunit/workunit.cpp:
```diff
@@ -2657,85 +2660,192 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+    }
+
+    loadGlobalAggregates(workunit);
+    if (isEmptyString(graphName))
```
It was originally, if (*graphName) and that was tested.  However, at the last minute I changed the code to use isEmptyString.  I know I shouldn't have done that - I should have tested every changed, even seemingly minor non-functional change.

jakesmith commented on system/jlib/jstats.cpp:
```diff
@@ -1994,12 +2098,95 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
+        const StatisticsMapping & mapping = totals.queryMapping();
+        bool updated = false;
+        // if this scope is not dirty, the aggregates are accurate at this level so return totals (no need to descend)
+        // Also if at sg scope, do not descend as aggregates does not need to be generated from below sg level
```
trivial: "..as aggregates do not need.."

shamser commented on ecl/eclagent/eclgraph.cpp:
```diff
@@ -927,6 +923,11 @@ void EclSubGraph::updateProgress(IStatisticGatherer &progress)
+    Owned<IStatisticCollection> statsCollection = progress.getResult();
+    const cost_type costDiskAccess = aggregateStatistic(StCostFileAccess, statsCollection);
+    if (costDiskAccess)
+        progress.addStatistic(StCostFileAccess, costDiskAccess);
```
This is existing relocated from EclSubGraph::updateProgress() line 900.  I hadn't considered changing how it was working.  

I'll modify to add a member function to IStatisticCollection and use that.

jakesmith commented on common/workunit/workunit.cpp:
```diff
@@ -2657,85 +2660,192 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+            case SSTgraph:
+                for (unsigned i=0; i<numStats; ++i)
+                {
+                    StatisticKind kind = aggregateKindsMapping.getKind(i);
```
Related to refreshAggregates / isDirty comments, and since regressAggretates is always done in order to push updates, should both these recursive descents into scopes to check for 'dirty' be combined?
i.e. so that updateAggregates only recalculates aggregates it needs to (if the stat is dirty) and writes them at that point (but commits to Dali at end).

shamser replied:
I've combined the refreshAggregates with updateAggregates so that now refreshAggregate can use a callback function to update an modified aggregate.

shamser commented on common/workunit/workunit.cpp:
```diff
@@ -2657,85 +2660,192 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+                MemoryBuffer serialized;
+                decompressToBuffer(serialized, compressed);
+
+                unsigned version;
```
I've removed the code to load sg stats in global collection- this entire section removed.

shamser commented on common/workunit/workunit.cpp:
```diff
@@ -2657,85 +2660,192 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+        Linked<IStatisticCollection> statsCollection;
+    };
+
+    WuScopeFilter filter;
```
@ghalliday You were going to come back to do this.  Any ideas on making this more efficient?

ghalliday replied:
Yes it could be made more efficient by directly walking the stats list.  If this works I wouldn't change it for this PR though - something to return to and simplify the code.

shamser commented on common/workunit/workunit.cpp:
```diff
@@ -2657,85 +2660,192 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
 
-void updateSpillSize(IWorkUnit * wu, const char * scope, StatisticScopeType scopeType)
+// Prune all subgraph descendent stats (leaving subgraph stats for future aggregation)
+void GlobalStatisticCollection::pruneSubGraphDescendants(unsigned wfid, const char *graphName, unsigned sgId)
```
There was couple of reasons for implementing the functionality like this
1) to avoid having to copy stats between the gatherer and collector, especially when the updates can be frequent (when lots of small graphs execute).
2) to be able cache stats so that in the future stats may be updated at fixed intervals.  At the moment when many small graphs executed, the stats (GraphProgress) much more frequently than the fixed heartbeat interval defined in the demonserver.  (Updates to any number of sg to GraphProgress may be made in one go.)

I can leave the existing gatherer code as is and merge the necessary stats from the gatherer into the GlobalAggregator.  Let me know if you think we should do this.

ghalliday replied:
1. Copying the stats is trivial.  You are copying ~4 stats from one structure to another.
2. That can be solved in a different way.  Saving a queue of blobs to commit.

Let's arrange a walkthrough to discuss before heading off an making any changes though.

shamser committed:
HPCC-29657 Change following review

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.com>

shamser commented:
@ghalliday @jakesmith I have addressed most of the review comments.  However, I haven't addressed keeping the existing gatherer code and having a separate Aggregator object.  I think the reason for the current design of using a gatherer to write director to the global stats object, is that I was
1) trying to avoid copying values, 
2) provide the possibility of some improvements in the future (such as caching sg stats and writing them out in batches)
3) possibly generate aggregates from sg activities, removing the need of having sg level aggregates scattered throughout the graph code.

ghalliday commented on system/jlib/jstats.h:
```diff
@@ -883,8 +912,11 @@ extern jlib_decl StatisticCreatorType queryCreatorType(const char * sct, Statist
 
 extern jlib_decl IStatisticGatherer * createStatisticsGatherer(StatisticCreatorType creatorType, const char * creator, const StatsScopeId & rootScope);
+extern jlib_decl IStatisticGatherer * createStatisticsGatherer(IStatisticCollection * stats);
+extern jlib_decl IStatisticCollection * createRootStatisticCollection(StatisticCreatorType creatorType, const char * creator, const StatsScopeId & rootScope, const StatsScopeId & graphScope, IStatisticCollection * sgCollection=nullptr);
```
This looks wrong.  A **Root**StatisticCollection is meant to have no parent.  I think this is another result of mixing up gathering the statisics with calculating aggregates.

ghalliday commented on common/workunit/workunit.cpp:
```diff
@@ -2657,85 +2660,104 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+    // Marking the collection dirty here is not ideal.  It would be better to have a call to IStatisticCollection::setStatistic mark the scope as dirty.
+    // However, this would be inefficient as each call to IStatisticCollection::setStatistic would require the dirty flag to be set for all parent scopes.
+    sgScopeCollection->markDirty();
+    return createRootStatisticCollection(creatorType, creator, wfScopeId, graphScopeId, sgScopeCollection);
```
This is confusing.  It is adding a "Root" scope with an id of wfid underneath a scope of workflow/graph/subgraph.  So effectively in the collection it is global->wfid->graph->subgraph->wfid->.
It make me more convinced it is a mistake to try and merge these two concepts.
It also means roxie works rather differently from eclagent/thor.

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -2190,49 +2350,59 @@ IStatisticCollection * createStatisticCollection(MemoryBuffer & in)
     {
         StatsScopeId scopeId(SSTchannel, id);
-        CStatisticCollection & tos = scopes.tos();
+        IStatisticCollection & tos = scopes.tos();
```
Still lots of unrequired changes from CStatisticCollection to IStatisticCollection throught the file that complicate the compare.

ghalliday commented on system/jlib/jstats.cpp:
```diff
@@ -1994,12 +2086,121 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
+                        if (value || includeStatisticIfZero(kind))
+                        {
+                            StringBuffer s;
+                            (fWhenAggregateUpdated)(getFullScope(s).str(), queryScopeType(), kind, value);
```
trivial.  I don't think the brackets are unneeded and confusing.

ghalliday reviewed (commented):
@shamser I think this is looking better.  I am still not convinced by doubling up the structure for two purposes - but let's discuss before making any changes.
